File: iphone/Classes/TiStylesheet.m
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2009-2010 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */

#import "TiStylesheet.h"
#import "TiBase.h"

#define DEBUG_STYLESHEETS 0

@implementation TiStylesheet

- (id)init
{
  if ((self = [super init])) {
    NSString *mainBundlePath = [[NSBundle mainBundle] bundlePath];
    NSString *plistPath = [mainBundlePath stringByAppendingPathComponent:@"stylesheet.plist"];
    DebugLog(@"[DEBUG] Reading stylesheet from: %@", plistPath);
    NSDictionary *dictionary = [[NSDictionary alloc] initWithContentsOfFile:plistPath];
    classesDict = [[dictionary objectForKey:@"classes"] retain];
    classesDictByDensity = [[dictionary objectForKey:@"classes_density"] retain];
    idsDict = [[dictionary objectForKey:@"ids"] retain];
    idsDictByDensity = [[dictionary objectForKey:@"ids_density"] retain];
    tagsDict = [[dictionary objectForKey:@"tags"] retain];
    tagsDictByDensity = [[dictionary objectForKey:@"tags_density"] retain];

#if defined(DEBUG) && DEBUG_STYLESHEETS == 1
    NSLog(@"[STYLESHEET DEBUG] ClassesDict = %@", classesDict);
    NSLog(@"[STYLESHEET DEBUG] ClassesDictByDensity = %@", classesDictByDensity);
    NSLog(@"[STYLESHEET DEBUG] IdsDict = %@", idsDict);
    NSLog(@"[STYLESHEET DEBUG] IdsDictByDensity = %@", idsDictByDensity);
#endif
    [dictionary release];
  }
  return self;
}

- (void)dealloc
{
  RELEASE_TO_NIL(classesDict);
  RELEASE_TO_NIL(classesDictByDensity);
  RELEASE_TO_NIL(idsDict);
  RELEASE_TO_NIL(idsDictByDensity);
  RELEASE_TO_NIL(tagsDict);
  RELEASE_TO_NIL(tagsDictByDensity);
  [super dealloc];
}

- (id)stylesheet:(NSString *)objectId density:(NSString *)density basename:(NSString *)basename classes:(NSArray *)classes tags:(NSArray *)tags
{
#if DEBUG_STYLESHEETS == 1
  NSLog(@"[STYLESHEET DEBUG] stylesheet -> objectId: %@, density: %@, basename: %@", objectId, density, basename);
  for (int i = 0; i < [classes count]; i++) {
    NSLog(@"[STYLESHEET DEBUG] -> class: %@", [classes objectAtIndex:i]);
  }

  for (int i = 0; i < [tags count]; i++) {
    NSLog(@"[STYLESHEET DEBUG] -> tag: %@", [tags objectAtIndex:i]);
  }
#endif

  /*
	 CSS selector priority order (lowest to highest) is
	 - Tag selectors
	 - Classes
	 - ID selectors
	*/

  NSMutableDictionary *result = [NSMutableDictionary dictionary];
  if (tags != nil) {
    NSEnumerator *tagEnum = [tags objectEnumerator];
    id tagName;

    while (tagName = [tagEnum nextObject]) {
      NSDictionary *tags = [[tagsDict objectForKey:basename] objectForKey:tagName],
                   *tagsD = [[[tagsDictByDensity objectForKey:basename] objectForKey:density] objectForKey:tagName];

      if (tags != nil) {
        [result addEntriesFromDictionary:tags];
      }
      if (tagsD != nil) {
        [result addEntriesFromDictionary:tagsD];
      }
    }
  }

  NSEnumerator *classEnum = [classes objectEnumerator];
  id className;
  while (className = [classEnum nextObject]) {
    NSDictionary *classes = [[classesDict objectForKey:basename] objectForKey:className];
    NSDictionary *classesD = [[[classesDictByDensity objectForKey:basename] objectForKey:density] objectForKey:className];
    if (classes != nil) {
      [result addEntriesFromDictionary:classes];
    }
    if (classesD != nil) {
      [result addEntriesFromDictionary:classesD];
    }
  }

  NSDictionary *ids = [[idsDict objectForKey:basename] objectForKey:objectId];
  NSDictionary *idsD = [[[idsDictByDensity objectForKey:basename] objectForKey:density] objectForKey:objectId];
  if (ids != nil) {
    [result addEntriesFromDictionary:ids];
  }
  if (idsD != nil) {
    [result addEntriesFromDictionary:idsD];
  }

#if DEBUG_STYLESHEETS == 1
  NSLog(@"[STYLESHEET DEBUG] stylesheet -> %@", result);
#endif
  return result;
}

- (BOOL)basename:(NSString *)basename density:(NSString *)density hasTag:(NSString *)tagName
{
  return ([[tagsDict objectForKey:basename] objectForKey:tagName] != nil ||
      [[[tagsDictByDensity objectForKey:basename] objectForKey:density] objectForKey:tagName]);
}

@end

File: android/titanium/src/java/org/appcelerator/titanium/TiStylesheet.java
/**
 * Appcelerator Titanium Mobile
 * Copyright (c) 2011-2012 by Appcelerator, Inc. All Rights Reserved.
 * Licensed under the terms of the Apache Public License
 * Please see the LICENSE included with this distribution for details.
 */
package org.appcelerator.titanium;

import java.util.Collection;
import java.util.HashMap;

import org.appcelerator.kroll.KrollDict;
import org.appcelerator.kroll.common.Log;

public abstract class TiStylesheet {
	private static final String TAG = "TiStylesheet";
	
	protected final HashMap<String,HashMap<String,KrollDict>> classesMap;
	protected final HashMap<String,HashMap<String,KrollDict>> idsMap;
	protected final HashMap<String,HashMap<String,HashMap<String,KrollDict>>> classesDensityMap;
	protected final HashMap<String,HashMap<String,HashMap<String,KrollDict>>> idsDensityMap;
	
	// The concrete implementation fills these
	public TiStylesheet() {
		classesMap = new HashMap<String,HashMap<String,KrollDict>>();
		idsMap = new HashMap<String,HashMap<String,KrollDict>>();
		classesDensityMap = new HashMap<String, HashMap<String,HashMap<String,KrollDict>>>();
		idsDensityMap = new HashMap<String, HashMap<String,HashMap<String,KrollDict>>>();
	}

	protected void addAll(KrollDict result, HashMap<String, KrollDict> map, String key) {
		if (map != null) {
			KrollDict d = map.get(key);
			if (d != null) {
				result.putAll(d);
			}
		}
	}
	
	public final KrollDict getStylesheet(String objectId, Collection<String> classes, String density, String basename)
	{
		Log.d(TAG, "getStylesheet id: " + objectId + ", classes: " + classes + ", density: " + density + ", basename: "
			+ basename, Log.DEBUG_MODE);
		
		KrollDict result = new KrollDict();
		if (classesMap != null)
		{
			HashMap<String, KrollDict> classMap = classesMap.get(basename);
			HashMap<String, KrollDict> globalMap = classesMap.get("global");
			if (globalMap != null || classMap != null) {
				for (String clazz : classes) {
					addAll(result, globalMap, clazz);
					addAll(result, classMap, clazz);
				}
			}
		}
		if (classesDensityMap != null)
		{
			HashMap<String,KrollDict> globalDensityMap = null;
			if (classesDensityMap.containsKey("global")) {
				globalDensityMap = classesDensityMap.get("global").get(density);
			}

			HashMap<String, KrollDict> classDensityMap = null;
			if (classesDensityMap.containsKey(basename)) {
				classDensityMap = classesDensityMap.get(basename).get(density);
			}

			if (globalDensityMap != null || classDensityMap != null) {
				for (String clazz : classes) {
					addAll(result, globalDensityMap, clazz);
					addAll(result, classDensityMap, clazz);
				}
			}
		}
		if (idsMap != null && objectId != null)
		{
			addAll(result, idsMap.get("global"), objectId);
			addAll(result, idsMap.get(basename), objectId);
		}
		if (idsDensityMap != null && objectId != null)
		{
			HashMap<String,KrollDict> globalDensityMap = null;
			if (idsDensityMap.containsKey("global")) {
				globalDensityMap = idsDensityMap.get("global").get(density);
			}
			HashMap<String,KrollDict> idDensityMap = null;
			if (idsDensityMap.containsKey(basename)) {
				idDensityMap = idsDensityMap.get(basename).get(density);
			}
			addAll(result, globalDensityMap, objectId);
			addAll(result, idDensityMap, objectId);
		}
		return result;
	}
}
